File: src/main/java/nl/jhvh/java/geometry/model/Dimensional.java
package nl.jhvh.java.geometry.model;

import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.Map;

import static java.util.stream.Collectors.toMap;

public enum Dimensional {
    oneDimensional(1),
    twoDimensional(2),
    threeDimensional(3);

    // dimensionCount must be unique
    // If not unique, mapping would fail with ExceptionInInitializerError - and that's what we want anyway !
    // Caused by: java.lang.IllegalStateException: Duplicate key .. (attempted merging values ... and ...)
    private static final Map<Integer, Dimensional> dimensionMapper = Arrays.stream(values())
            .collect(toMap(dm -> dm.dimensionCount, dm -> dm));

    public final int dimensionCount;

    Dimensional(int dimensionCount) {
        this.dimensionCount = dimensionCount;
    }

    @Nullable
    public static Dimensional byDimensionCount(int dimensionCount) {
        return dimensionMapper.get(dimensionCount);
    }
}

File: src/main/kotlin/nl/jhvh/kotlin/geometry/model/Dimensional.kt
package nl.jhvh.kotlin.geometry.model

enum class Dimensional(val dimensionCount: Int) {
    oneDimensional(1),
    twoDimensional(2),
    threeDimensional(3);

    companion object Mapping {
        // dimensionCount must be unique, but this mapping does NOT throw an exception when not unique!
        // If we really want so, we have to check uniqueness by ourselves
        private val dimensionalMapper: Map<Int, Dimensional> = values().map { it.dimensionCount to it }.toMap()

        init {
            // Throws IllegalStateException when not meeting condition
            check(dimensionalMapper.size == values().size) {"Duplicate key in ${values().map { "${it.dimensionCount} -> $it" } }"}
        }

        fun byDimensionCount(dimensionCount: Int): Dimensional? = dimensionalMapper[dimensionCount]
    }
}
